### ISSUE TITLE
Intrinsic is(Data)(Type) API for `data`

### ISSUE THREAD

**Issue by benjamin-jin** [sifive/wake#843]
Given the following case,
```
data A = 
  B
  C (some : SomeType)
def mylist = (B, C 3, B, C 4, Nil)
```
to filter mylist with type C B only, we do as following
```
def isB : A => Boolean = match _
  B = True
  _ = False
def isC : A => Boolean = match _
  C _   = True
  _     = False

def mylist = (B, C 3, B, C 4, Nil)
def b_only = mylist | filter isB
def c_only = mylist | filter isC
```
This is totally fine when data A has few types.

However, time to time we may add new case of data A, and every time we need to define new `isD`, `isE`.

How about adding new API to check type of data instead?

**Comment by terpstra** [comment]
Yeah, analogous to how `tuple` has `get`/`set`/`edit` methods, I have considered adding automatic helper methods along the lines of `getAC: A => Option (some: SomeType)`. I just haven't decided on what form these should take. Furthermore, I remain hopeful that we can get field accessor methods like `obj.field` to work in the future, and I'd like these to line up nicely, whenever that happens.

**Comment by benjamin-jin** [comment]
Do you mean tuple instance's filed ?
